libmamba/src/core/package_fetcher.cpp

```diff
@@ -0,0 +1,428 @@
+        if (fs::exists(dest_dir))
+        {
+            fs::remove_all(dest_dir);
+        }
+    }
+
+    /*******************
+     * Private methods *
+     *******************/
+
+    const std::string& PackageFetcher::filename() const
+    {
+        return m_package_info.fn;
+    }
+
+    const std::string& PackageFetcher::url() const
+    {
+        return m_url;
+    }
+
+    const std::string& PackageFetcher::sha256() const
+    {
+        return m_package_info.sha256;
+    }
+
+    const std::string& PackageFetcher::md5() const
+    {
+        return m_package_info.md5;
+    }
+
+    std::size_t PackageFetcher::expected_size() const
+    {
+        return m_package_info.size;
+    }
+
+    auto PackageFetcher::validate_size(std::size_t downloaded_size) const -> ValidationResult
+    {
+        auto res = ValidationResult::VALID;
+        if (expected_size() && expected_size() != downloaded_size)
+        {
+            res = ValidationResult::SIZE_ERROR;
+            LOG_ERROR << "File not valid: file size doesn't match expectation " << m_tarball_path
+                      << "\nExpected: " << expected_size() << "\nActual: " << downloaded_size
+                      << "\n";
+            Console::instance().print(filename() + " tarball has incorrect size");
+            // TODO: terminate monitor
+        }
+        return res;
+    }
+
+    auto PackageFetcher::validate_checksum(CheckSumParams params) const -> ValidationResult
+    {
+        auto res = ValidationResult::VALID;
+        if (params.actual != params.expected)
+        {
+            res = params.error;
+            LOG_ERROR << "File not valid: " << params.name << " doesn't match expectation "
+                      << m_tarball_path << "\nExpected: " << params.expected
+                      << "\nActual: " << params.actual << "\n";
+            Console::instance().print(filename() + " tarball has incorrect " + params.name);
+            // TODO: terminate monitor
+        }
+        return res;
+    }
+
+    void PackageFetcher::write_repodata_record(const fs::u8path& base_path) const
+    {
+        fs::u8path repodata_record_path = base_path / "info" / "repodata_record.json";
+        fs::u8path index_path = base_path / "info" / "index.json";
+
+        nlohmann::json index, solvable_json;
+        std::ifstream index_file = open_ifstream(index_path);
+        index_file >> index;
+
+        solvable_json = m_package_info.json_record();
+        index.insert(solvable_json.cbegin(), solvable_json.cend());
+
+        if (index.find("size") == index.end() || index["size"] == 0)
+        {
+            index["size"] = fs::file_size(m_tarball_path);
+        }
+
+        std::ofstream repodata_record(repodata_record_path.std_path());
+        repodata_record << index.dump(4);
+    }
+
+    namespace
+    {
+        std::mutex urls_txt_mutex;
+    }
+
+    void PackageFetcher::update_urls_txt() const
+    {
+        std::lock_guard<std::mutex> lock(urls_txt_mutex);
```

Comment:
I'm not sure that's necessary? Isnt `ofstream` already writing in 1 locked operation if there is only one value to write?

Reply:
I have no idea. I'll add a comment so we can check it later, I just took the previous code as is.